// @amityco/ts-sdk@0.0.1-beta.40/src/client/api/getToken.ts
import { getActiveClient } from './activeClient';

import { getDeviceInfo } from '~/core/device';

/**
 * Retrieves a pair of {@link Amity.Tokens} necessary for connection
 *
 * @param params.userId The userId to use to issue a token
 * @param params.displayName The user's displayName
 * @param params.deviceId The user's device Id (can be manually set for native users)
 * @param params.authToken The authentication token - necessary when network option is set to secure
 * @return The accessToken for the given userId
 *
 * @category Client API
 * @hidden
 */
export const getToken = async (params: {
  userId: Amity.User['userId'];
  displayName: Amity.User['displayName'];
  authToken?: string;
  deviceId: Amity.Device['deviceId'];
}) => {
  const client = getActiveClient();
  client.log('client/api/getToken', params);

  const deviceInfo = getDeviceInfo();

  const { data } = await client.http.post<Amity.Tokens & { users: Amity.User[] }>(
    '/api/v3/sessions',
    {
      ...params,
      deviceInfo,
    },
    {
      // @ts-ignore
      headers: { 'X-API-Key': client.apiKey },
    },
  );

  return data;
};
