// @amityco/ts-sdk@0.0.1-beta.40/src/post/api/deletePost.ts
import { getActiveClient } from '~/client/api';

import { fireEvent } from '~/core/events';
import { upsertInCache } from '~/cache/api';
import { getCommunity } from '~/community/api';
import { pushToTombstone } from '~/cache/api/pushToTombstone';

import { getPost } from './getPost';

/**
 * ```js
 * import { deletePost } from '@amityco/ts-sdk'
 * const success = await deletePost('foobar')
 * ```
 *
 * Deletes a {@link Amity.Post}
 *
 * @param postId The {@link Amity.Post} ID to delete
 * @return A success boolean if the {@link Amity.Post} was deleted
 *
 * @category Post API
 * @async
 */
export const deletePost = async (
  postId: Amity.Post['postId'],
  permanent = false,
): Promise<Amity.Post> => {
  const client = getActiveClient();
  client.log('post/deletePost', postId);

  const post = await getPost(postId);

  // API-FIX: This endpoint has not been implemented yet.
  await client.http.delete<{ success: boolean }>(`/api/v4/posts/${encodeURIComponent(postId)}`, {
    params: {
      postId,
      permanent,
    },
  });

  // there is currently a limitation which doesn't allow us to fire event to tell that community
  // has been updated. reason is that, when the object is deleted, we don't have its `communityId`
  // and so we cannot refetch the community or advertise on events. hopefully this should be solved
  // later when realtime events covers that for us.
  if (post.data.targetType === 'community') {
    const community = await getCommunity(post.data.targetId);
    fireEvent('community.updated', {
      communities: [community.data],
      categories: [],
      communityUsers: [],
      feeds: [],
      files: [],
      users: [],
    });
  }

  // to support hard deletion
  const deleted = { ...post.data, isDeleted: true };

  fireEvent('post.deleted', {
    posts: [deleted],
    categories: [],
    comments: [],
    communities: [],
    communityUsers: [],
    feeds: [],
    files: [],
    postChildren: [],
    users: [],
  });

  if (permanent) {
    setTimeout(() => {
      pushToTombstone('post', postId);
    }, 0);
  } else {
    upsertInCache(['post', 'get', postId], { isDeleted: true });
  }

  return deleted;
};
