// @amityco/ts-sdk@0.0.1-beta.40/src/post/api/updatePost.ts
import { getActiveClient } from '~/client/api';

import { ingestInCache } from '~/cache/api/ingestInCache';
import { fireEvent } from '~/core/events';
import { prepareMembershipPayload } from '~/group/utils';

/**
 * ```js
 * import { updatePost } from '@amityco/ts-sdk'
 * const updated = await updatePost(postId, {
 *   data: { text: 'hello world' }
 * })
 * ```
 *
 * Updates an {@link Amity.Post}
 *
 * @param postId The ID of the {@link Amity.Post} to edit
 * @param patch The patch data to apply
 * @returns the updated {@link Amity.Post} object
 *
 * @category Post API
 * @async
 */
export const updatePost = async (
  postId: Amity.Post['postId'],
  patch: Patch<Amity.Post, 'data' | 'metadata' | 'mentionees' | 'tags'> & {
    attachments?: {
      type: 'image' | 'file' | 'video';
      fileId: Amity.File['fileId'];
    }[];
  },
): Promise<Amity.Cached<Amity.Post>> => {
  const client = getActiveClient();
  client.log('user/updatePost', patch);

  const { data: payload } = await client.http.put<Amity.PostPayload>(
    `/api/v4/posts/${encodeURIComponent(postId)}`,
    patch,
  );

  const data = prepareMembershipPayload(payload, 'communityUsers');

  const cachedAt = client.cache && Date.now();
  if (client.cache) ingestInCache(data, { cachedAt });

  fireEvent('post.updated', data);

  const { posts } = data;

  return {
    data: posts.find(post => post.postId === postId)!,
    cachedAt,
  };
};
