// @amityco/ts-sdk@0.0.1-beta.40/src/utils/tests/dummy/category.ts
export const category11 = {
  name: 'a-hello',
  updatedAt: '2022-11-25T04:19:34.446Z',
  createdAt: '2022-11-25T04:19:34.446Z',
  isDeleted: false,
  categoryId: '2ddea5ad242d0a2bd5dc5be8ae9fc8c2',
};

export const category12 = {
  name: 'b-testing-category',
  updatedAt: '2022-11-25T03:56:06.238Z',
  createdAt: '2022-11-25T03:56:06.238Z',
  isDeleted: false,
  categoryId: '88d175da65cbfeb39b81e722b322430e',
};

export const category13 = {
  name: 'c-deleted-category',
  updatedAt: '2022-11-25T03:56:06.238Z',
  createdAt: '2022-11-25T03:56:06.238Z',
  isDeleted: true,
  categoryId: '88d175da65cbfeb39b81e722b322430d',
};

export const category21 = {
  name: 'a-0asd',
  updatedAt: '2022-11-25T04:18:59.180Z',
  createdAt: '2022-11-25T04:18:59.180Z',
  isDeleted: false,
  categoryId: '10051d82a1f371c6be3a8f4a4654b8a1',
};

export const category22 = {
  name: 'b-A Category',
  updatedAt: '2022-11-25T03:56:29.037Z',
  createdAt: '2022-11-25T03:56:29.037Z',
  isDeleted: false,
  categoryId: '639942c6b2f121bed3901cf4a4c296e7',
};

export const categoryQueryResponse = {
  data: {
    categories: [category11, category12, category13],
    paging: {
      next: 'eyJza2lwIjoxMCwibGltaXQiOjEwfQ==',
    },
  },
};

export const categoryQueryResponsePage2 = {
  data: {
    categories: [category21, category22],
    paging: {
      next: 'eyJza2lwIjoxMCwibGltaXQiOjEwfQ==',
    },
  },
};
