// @amityco/ts-sdk@0.0.1-beta.40/src/utils/tests/client.ts
import {
  createClient,
  connectClient as sdkConnectClient,
  disconnectClient as sdkDisconnectClient,
} from '~/client/api';

import { user11 } from './dummy';

const sessionHandler: Amity.SessionHandler = {
  sessionWillRenewAccessToken(_) {
    // do nothing
  },
};

export const client = createClient('key', 'sg');
client.accessTokenExpiryWatcher = (_, __, ___) => () => {
  //
};

export const connectClient = async () => {
  const now = new Date();
  const nextMonth = new Date();
  nextMonth.setDate(now.getDate() + 30);

  client.mqtt.connect = jest.fn();
  client.mqtt.subscribe = jest.fn();
  client.http.post = jest.fn().mockReturnValueOnce({
    data: {
      accessToken: 'accessToken',
      refreshToken: 'refreshToken',
      users: [user11],
      expiresAt: nextMonth.toISOString(),
      issuedAt: now.toISOString(),
    },
  });

  setTimeout(() => {
    client.ws.emit('connect');
    // fixes jest warning of open handles
  }, 500).unref();

  await sdkConnectClient({ userId: 'test' }, sessionHandler);
};

export const disconnectClient = async () => {
  sdkDisconnectClient().then(() => client.ws.emit('disconnect'));
};

export default {};
