// @amityco/ts-sdk@0.0.1-beta.40/src/user/observers/liveUsers.ts
/* eslint-disable no-use-before-define */
import { getResolver } from '~/core/model';
import { pullFromCache, pushToCache } from '~/cache/api';
import { getActiveClient } from '~/client/api';

import {
  createQuery,
  runQuery,
  queryOptions,
  filterByPropEquality,
  sortByLastCreated,
  sortByDisplayName,
  sortByFirstCreated,
} from '~/core/query';

import {
  COLLECTION_DEFAULT_CACHING_POLICY,
  COLLECTION_DEFAULT_PAGINATION_LIMIT,
  ENABLE_CACHE_MESSAGE,
} from '~/utils/constants';

import { queryUsers } from '../api';
import {
  onUserFlagged,
  onUserUpdated,
  onUserUnflagged,
  onUserFlagCleared,
  onUserDeleted,
} from '../events';

/**
 * ```js
 * import { liveUsers } from '@amityco/ts-sdk'
 *
 * let users = []
 * const unsub = liveUsers({}, response => merge(users, response.data))
 * ```
 *
 * Observe all mutations on a list of {@link Amity.User}s
 *
 * @param params for querying users
 * @param callback the function to call when new data are available
 * @returns An {@link Amity.Unsubscriber} function to run when willing to stop observing the users
 *
 * @category Category Live Collection
 */
export const liveUsers = (
  params: Amity.UserLiveCollection,
  callback: Amity.LiveCollectionCallback<Amity.User>,
  config?: Amity.LiveCollectionConfig,
) => {
  const { log, cache } = getActiveClient();

  if (!cache) {
    console.log(ENABLE_CACHE_MESSAGE);
  }

  const timestamp = Date.now();
  log(`liveUsers(tmpid: ${timestamp}) > listen`);

  const { limit: queryLimit, ...queryParams } = params;

  const limit = queryLimit ?? COLLECTION_DEFAULT_PAGINATION_LIMIT;
  const { policy = COLLECTION_DEFAULT_CACHING_POLICY } = config ?? {};

  const disposers: Amity.Unsubscriber[] = [];
  const cacheKey = ['user', 'collection', {}];

  const responder = (data: Amity.UserLiveCollectionCache) => {
    let users: Amity.User[] =
      data.data
        .map(userId => pullFromCache<Amity.User>(['user', 'get', userId])!)
        .filter(Boolean)
        .map(({ data }) => data) ?? [];

    users = filterByPropEquality(users, 'displayName', params.displayName);

    switch (params.sortBy) {
      case 'firstCreated':
        users = users.sort(sortByFirstCreated);
        break;

      case 'lastCreated':
        users = users.sort(sortByLastCreated);
        break;

      default:
        users = users
          // this needs to be aligned with the backend data type
          .map(u => (u.displayName ? u : { ...u, displayName: '' }))
          // @ts-ignore
          .sort(sortByDisplayName);
    }

    callback({
      onNextPage: onFetch,
      data: users,
      hasNextPage: !!data.params?.page,
      loading: data.loading,
      error: data.error,
    });
  };

  const realtimeRouter = (_: Amity.UserActionType) => (user: Amity.User) => {
    const collection = pullFromCache<Amity.CommunityLiveCollectionCache>(cacheKey)?.data;
    if (!collection) return;

    collection.data = [...new Set([user.userId, ...collection.data])];

    pushToCache(cacheKey, collection);
    responder(collection);
  };

  const onFetch = () => {
    const collection = pullFromCache<Amity.UserLiveCollectionCache>(cacheKey)?.data;

    const users = collection?.data ?? [];

    if (users.length > 0 && !collection?.params.page) return;

    const query = createQuery(queryUsers, {
      ...queryParams,
      page: collection?.params.page ?? { limit },
    });

    runQuery(
      query,
      ({ data: result, error, loading, nextPage: page }) => {
        const data = {
          loading,
          error,
          params: { page },
          data: users,
        };

        if (result) {
          data.data = [...new Set([...users, ...result.map(getResolver('user'))])];
        }

        pushToCache(cacheKey, data);

        responder(data);
      },
      queryOptions(policy),
    );
  };

  disposers.push(
    onUserUpdated(realtimeRouter('onUpdate')),
    onUserDeleted(realtimeRouter('onDelete')),
    onUserFlagged(realtimeRouter('onFlagged')),
    onUserUnflagged(realtimeRouter('onUnflagged')),
    onUserFlagCleared(realtimeRouter('onFlagCleared')),
  );

  onFetch();

  return () => {
    log(`liveUsers(tmpid: ${timestamp}) > dispose`);
    disposers.forEach(fn => fn());
  };
};
