// @amityco/ts-sdk@0.0.1-beta.40/src/reaction/events/onReactorAdded.ts
import { getActiveClient } from '~/client/api';
import { ingestInCache } from '~/cache/api/ingestInCache';
import { createEventSubscriber } from '~/core/events';
import { prepareMessagePayload } from '~/message/utils';

/**
 * ```js
 * import { onReactorAdded } from '@amityco/ts-sdk'
 * const dispose = onReactorAdded('post', postId, reactor => {
 *   // ...
 * })
 * ```
 *
 * Fired when an {@link Amity.Reactor} has been added
 *
 * @param {@link Amity.ReactableType} referenceType
 * @param {string} referenceId
 * @param callback The function to call when the event was fired
 * @returns an {@link Amity.Unsubscriber} function to stop listening
 *
 * @category Events
 * */
export const onReactorAdded = (
  referenceType: Amity.ReactableType,
  referenceId: Amity.Reaction['referenceId'],
  callback: Amity.Listener<Amity.Reactor>,
): Amity.Unsubscriber => {
  const client = getActiveClient();

  const callbackWrapper = (
    referenceType_: Amity.ReactableType,
    referenceId_: Amity.Reaction['referenceId'],
    reaction: Amity.Reactor,
  ) => {
    if (referenceType_ === referenceType && referenceId_ === referenceId) {
      callback(reaction);
    }
  };

  if (referenceType === 'message') {
    const filter = (rawPayload: Amity.MessagePayload) => {
      const payload = prepareMessagePayload(rawPayload);
      if (!payload.messages[0].latestReaction) return;

      const { eventName, referenceId, referenceType, userDisplayName, ...reactor } =
        payload.messages[0].latestReaction;

      if (eventName !== 'add') return;

      ingestInCache(payload);
      // FIXME: correct reactions type in model
      // @ts-ignore
      ingestInCache({ reactions: [reactor as Amity.Reactor] });

      callbackWrapper('message', payload.messages[0].messageId, reactor);
    };

    return createEventSubscriber(client, 'reaction/onReactorAdded', 'message.updated', filter);
  }

  if (referenceType === 'post') {
    const filter = (payload: Amity.PostPayload & { reactor: Amity.Reactor }) => {
      const { reactor, ...rest } = payload;

      ingestInCache(rest as Amity.ProcessedPostPayload);
      // FIXME: correct reactions type in model
      // @ts-ignore
      ingestInCache({ reactions: [reactor] });

      callbackWrapper('post', payload.posts[0].postId, reactor);
    };

    return createEventSubscriber(client, 'post.addReaction', 'post.addReaction', filter);
  }

  const filter = (payload: Amity.CommentPayload & { reactor: Amity.Reactor }) => {
    const { reactor, ...rest } = payload;

    ingestInCache(rest as Amity.CommentPayload);
    // FIXME: correct reactions type in model
    // @ts-ignore
    ingestInCache({ reactions: [reactor] });

    callbackWrapper('comment', payload.comments[0].commentId, reactor);
  };

  return createEventSubscriber(client, 'comment.addReaction', 'comment.addReaction', filter);
};
