// @amityco/ts-sdk@0.0.1-beta.40/src/comment/api/deleteComment.ts
import { getActiveClient } from '~/client/api';

import { upsertInCache } from '~/cache/api';
import { fireEvent } from '~/core/events';

import { getPost } from '~/post/api/getPost';
import { pushToTombstone } from '~/cache/api/pushToTombstone';
import { scheduleTask } from '~/core/microtasks';

import { getComment } from './getComment';

/**
 * ```js
 * import { deleteComment } from '@amityco/ts-sdk'
 * const success = await deleteComment('foobar')
 * ```
 *
 * Deletes a {@link Amity.Comment}
 *
 * @param commentId The {@link Amity.Comment} ID to delete
 * @return A success boolean if the {@link Amity.Comment} was deleted
 *
 * @category Comment API
 * @async
 */
export const deleteComment = async (
  commentId: Amity.Comment['commentId'],
  permanent = false,
): Promise<Amity.Comment> => {
  const client = getActiveClient();
  client.log('comment/deleteComment', commentId);

  const comment = await getComment(commentId);

  // API-FIX: This endpoint has not been implemented yet.
  await client.http.delete<{ success: boolean }>(
    `/api/v4/comments/${encodeURIComponent(commentId)}`,
    {
      params: {
        commentId,
        permanent,
      },
    },
  );

  // to support hard deletion
  const deleted = { ...comment.data, isDeleted: true };

  const post = await getPost(comment.data.referenceId);

  fireEvent('post.updated', {
    posts: [post.data],
    categories: [],
    comments: [],
    communities: [],
    communityUsers: [],
    feeds: [],
    files: [],
    postChildren: [],
    users: [],
  });

  fireEvent('comment.deleted', {
    comments: [deleted],
    commentChildren: [],
    files: [],
    users: [],
  });

  if (permanent) {
    scheduleTask(() => pushToTombstone('comment', commentId));
  } else {
    upsertInCache(['comment', 'get', commentId], { isDeleted: true });
  }

  return deleted;
};
