// @amityco/ts-sdk@0.0.1-beta.40/src/core/model/index.ts
export * from './idResolvers';
export * from './identifyModel';

/**
 * A map of v3 response keys to a store name.
 * @hidden
 */
export const PAYLOAD2MODEL: Record<string, Amity.Domain> = {
  users: 'user',
  files: 'file',
  roles: 'role',

  channels: 'channel',
  messageFeeds: 'subChannel',
  channelUsers: 'channelUsers',
  messages: 'message',

  communities: 'community',
  categories: 'category',
  communityUsers: 'communityUsers',
  posts: 'post',
  postChildren: 'post',
  comments: 'comment',
  polls: 'poll',
  reactions: 'reaction',

  videoStreamings: 'stream',

  follows: 'follow',
  followCounts: 'followCount',

  feeds: 'feed',
};

/** hidden */
export const hasUpdates = (prevData: Amity.UpdatedAt, nextData: Amity.UpdatedAt): boolean => {
  return new Date(prevData.updatedAt!) < new Date(nextData.updatedAt!);
};

/** hidden */
export function getFutureDate(date: string | undefined = new Date().toISOString()): string {
  return new Date(new Date(date).getTime() + 1).toISOString();
}

export function getPastDate(date: string | undefined = new Date().toISOString()): string {
  return new Date(new Date(date).getTime() - 1).toISOString();
}
